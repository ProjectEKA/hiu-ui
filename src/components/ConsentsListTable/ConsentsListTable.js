import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import * as PropTypes from 'prop-types';
import ArrowForwardIosIcon from '@material-ui/icons/ArrowForwardIos';
import MaterialTable from 'material-table';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import { formatDateString } from '../common/HealthInfo/FhirResourcesUtils';
import compareDates from '../common/DateUtil';

const useStyles = makeStyles(theme => ({
  table: {
    clear: 'both',
    backgroundColor: theme.palette.primary.main
  },
  title: {
    textTransform: 'uppercase'
  }
}));

const ConsentsListTable = ({ loadConsents, consentsList, loading }) => {
  const classes = useStyles();
  const [refreshCounter, setRefreshCounter] = useState(0);
  useEffect(() => {
    loadConsents();
  }, [refreshCounter]);

  const headerRow = {
    name: 'Name',
    jataayuId: 'Patient Identifier',
    requestStatus: 'Request Status',
    consentGrantedDate: 'Consent granted on',
    consentExpiryDate: 'Consent expiry on',
    consentCreatedDate: 'Consent created on'
  };

  function getStatusText(status) {
    switch (status.toUpperCase()) {
      case 'GRANTED':
        return 'Consent Granted';
      case 'POSTED':
        return 'Request Initiated';
      case 'ERRORED':
        return 'Request failed';
      case 'DENIED':
        return 'Request Denied';
      case 'REVOKED':
      return 'Consent Revoked';
      case 'EXPIRED':
      return 'Consent Expired';
      default:
        return 'Request sent';
    }
  }

  function isGranted(status) {
    return status.toUpperCase() == 'GRANTED';
  }

  function getPatientFullName(patient) {
    return `${patient.firstName} ${patient.lastName}`;
  }

  return (
    <div style={{ maxWidth: '100%' }}>
      <MaterialTable
        options={{
          headerStyle: {
            backgroundColor: '#009688',
            color: '#FFF'
          }
        }}
        isLoading={loading}
        className={classes.table}
        columns={[
          { title: headerRow.name, field: 'name' },
          { title: headerRow.jataayuId, field: 'id' },
          { title: headerRow.requestStatus, field: 'status' },
          {
            title: headerRow.consentCreatedDate,
            field: 'createdOn',
            customSort: (a, b) => compareDates(a.createdOn, b.createdOn)
          },
          {
            title: headerRow.consentGrantedDate,
            field: 'grantedOn',
            customSort: (a, b) => compareDates(a.grantedOn, b.grantedOn)
          },
          {
            title: headerRow.consentExpiryDate,
            field: 'expiredOn',
            customSort: (a, b) => compareDates(a.expiredOn, b.expiredOn)
          },
          { title: '', field: 'navLink', width: 50 }
        ]}
        data={consentsList
          .sort((a, b) => {
            const dateA = new Date(a.createdDate);
            const dateB = new Date(b.createdDate);
            return dateB - dateA;
          })
          .map(consent => ({
            name: getPatientFullName(consent.patient),
            id: consent.patient.id,
            status: getStatusText(consent.status),
            grantedOn: isGranted(consent.status)
              ? formatDateString(consent.approvedDate, true)
              : '-',
            expiredOn: isGranted(consent.status)
              ? formatDateString(consent.expiredDate, true)
              : '-',
            createdOn: formatDateString(consent.createdDate, true),
            navLink: isGranted(consent.status) ? (
              <Link to={`/health-info/${consent.consentRequestId}`}>
                <ArrowForwardIosIcon color="primary" />
              </Link>
            ) : (
              ''
            )
          }))}
        title={(
          <Typography className={classes.title} variant="h5">
            Consent List
          </Typography>
        )}
        actions={[
          {
            icon: 'refresh',
            tooltip: 'Refresh',
            isFreeAction: true,
            onClick: () => setRefreshCounter(refreshCounter + 1)
          }
        ]}
      />
    </div>
  );
};

const consentShape = PropTypes.shape({
  id: PropTypes.string,
  patient: PropTypes.shape({
    id: PropTypes.string,
    firstName: PropTypes.string,
    lastName: PropTypes.string
  }),
  status: PropTypes.string,
  approvedDate: PropTypes.string,
  expiredDate: PropTypes.string,
  createdDate: PropTypes.string
});

ConsentsListTable.propTypes = {
  loadConsents: PropTypes.func.isRequired,
  consentsList: PropTypes.arrayOf(consentShape).isRequired,
  loading: PropTypes.bool
};

ConsentsListTable.defaultProps = {
  loading: false
};

export default ConsentsListTable;
