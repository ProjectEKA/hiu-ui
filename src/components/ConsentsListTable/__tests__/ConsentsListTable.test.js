import React from 'react';
import { shallow } from 'enzyme';
import MaterialTable from 'material-table';
import ConsentsListTable from '../ConsentsListTable';

describe('ConsentsListTable', () => {
  const loadConsents = jest.fn();
  const mockData = [
    {
      patient: {
        firstName: 'John',
        lastName: 'Smith',
      },
      status: 'GRANTED',
      approvedDate: '2020-02-19T00:30:08.000+0000',
      expiredDate: '2020-02-20T03:17:48.603+0000',
      createdDate: '2020-02-18T03:17:48.603+0000',
      navLink: '',
    },
  ];
  global.TIMEZONE_OFFSET = '-05:00';
  const wrapper = shallow(
    <ConsentsListTable consentsList={mockData} loadConsents={loadConsents} theme="" />,
  );

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
    // expect(loadConsents).toHaveBeenCalled();
  });

  it('should render data in consent list when status is granted', () => {
    expect(wrapper.find(MaterialTable).length).toEqual(1);
    console.log(wrapper.find(MaterialTable).props().data[0].expiredOn);
    expect(wrapper.find(MaterialTable).props().data[0].name).toEqual('John Smith');
    expect(wrapper.find(MaterialTable).props().data[0].status).toEqual('Consent Granted');
    expect(wrapper.find(MaterialTable).props().data[0].grantedOn).toEqual('18/02/2020 07:30 PM');
    expect(wrapper.find(MaterialTable).props().data[0].expiredOn).toEqual('19/02/2020 10:17 PM');
    expect(wrapper.find(MaterialTable).props().data[0].createdOn).toEqual('17/02/2020 10:17 PM');
    expect(wrapper.find(MaterialTable).props().data[0].navLink).not.toEqual('');
  });

  it('should render data in consent list when status is not granted', () => {
    mockData[0].status = 'DENIED';
    wrapper.setProps({ consentsList: mockData });

    expect(wrapper.find(MaterialTable).length).toEqual(1);
    expect(wrapper.find(MaterialTable).props().data[0].name).toEqual('John Smith');
    expect(wrapper.find(MaterialTable).props().data[0].status).toEqual('Request Denied');
    expect(wrapper.find(MaterialTable).props().data[0].grantedOn).toEqual('-');
    expect(wrapper.find(MaterialTable).props().data[0].expiredOn).toEqual('-');
    expect(wrapper.find(MaterialTable).props().data[0].createdOn).toEqual('17/02/2020 10:17 PM');
    expect(wrapper.find(MaterialTable).props().data[0].navLink).toEqual('');
  });

  it('should have refresh actions', () => {
    expect(wrapper.find(MaterialTable).props().actions[0].icon).toEqual('refresh');
    expect(wrapper.find(MaterialTable).props().actions[0].isFreeAction).toEqual(true);
    expect(wrapper.find(MaterialTable).props().actions[0].tooltip).toEqual("Refresh");
    expect(wrapper.find(MaterialTable).props().actions[0].onClick).toBeDefined();
  });

  it('should have default props', () => {
    expect(ConsentsListTable.defaultProps.loading).toEqual(false);
  });
});
