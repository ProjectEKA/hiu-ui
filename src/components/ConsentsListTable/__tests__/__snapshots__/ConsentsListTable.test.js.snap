// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsentsListTable should render properly 1`] = `
<div
  style={
    Object {
      "maxWidth": "100%",
    }
  }
>
  <WithStyles(Component)
    actions={
      Array [
        Object {
          "icon": "refresh",
          "isFreeAction": true,
          "onClick": [Function],
          "tooltip": "Refresh",
        },
      ]
    }
    className="makeStyles-table-1"
    columns={
      Array [
        Object {
          "field": "name",
          "title": "Name",
        },
        Object {
          "field": "id",
          "title": "Patient Identifier",
        },
        Object {
          "field": "status",
          "title": "Request Status",
        },
        Object {
          "customSort": [Function],
          "field": "createdOn",
          "title": "Consent created on",
        },
        Object {
          "customSort": [Function],
          "field": "grantedOn",
          "title": "Consent granted on",
        },
        Object {
          "customSort": [Function],
          "field": "expiredOn",
          "title": "Consent expiry on",
        },
        Object {
          "field": "navLink",
          "title": "",
          "width": 50,
        },
      ]
    }
    data={
      Array [
        Object {
          "createdOn": "17/02/2020 10:17 PM",
          "expiredOn": "19/02/2020 10:17 PM",
          "grantedOn": "18/02/2020 07:30 PM",
          "id": undefined,
          "name": "John Smith",
          "navLink": <ForwardRef
            to="/health-info/undefined"
          >
            <UNDEFINED
              color="primary"
            />
          </ForwardRef>,
          "status": "Consent Granted",
        },
      ]
    }
    isLoading={false}
    options={
      Object {
        "headerStyle": Object {
          "backgroundColor": "#009688",
          "color": "#FFF",
        },
      }
    }
    title={
      <ForwardRef(WithStyles)
        className="makeStyles-title-2"
        variant="h5"
      >
        Consent List
      </ForwardRef(WithStyles)>
    }
  />
</div>
`;
